domains/transfer/Transfer.API/API/Claiming/Automation/ClaimService.cs

```diff
@@ -0,0 +1,79 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using API.Claiming.Api.Repositories;
+using API.Claiming.Automation.Services;
+using Microsoft.Extensions.Logging;
+using ProjectOrigin.WalletSystem.V1;
+
+namespace API.Claiming.Automation;
+
+public class ClaimService : IClaimService
+{
+    private readonly ILogger<ClaimService> logger;
+    private readonly IClaimRepository claimRepository;
+    private readonly IProjectOriginWalletService walletService;
+
+    public ClaimService(ILogger<ClaimService> logger, IClaimRepository claimRepository, IProjectOriginWalletService walletService)
+    {
+        this.logger = logger;
+        this.claimRepository = claimRepository;
+        this.walletService = walletService;
+    }
+
+    public async Task Run(CancellationToken stoppingToken)
+    {
+        while (!stoppingToken.IsCancellationRequested)
+        {
+            logger.LogInformation("ClaimService running at: {time}", DateTimeOffset.Now);
+            try
+            {
+                var claimSubjects = await claimRepository.GetClaimSubjects();
+                foreach (var subjectId in claimSubjects.Select(x => x.SubjectId))
+                {
+                    var certificates = await walletService.GetGranularCertificates(subjectId);
+                    var consumptionCertificates = certificates.Where(x => x.Type == GranularCertificateType.Consumption).ToList();
+                    var productionCertificates = certificates.Where(x => x.Type == GranularCertificateType.Production).ToList();
+
+                    await Claim(subjectId, consumptionCertificates, productionCertificates);
+                }
+            }
+            catch (Exception e)
+            {
+                logger.LogWarning("Something went wrong with the ClaimService: {exception}", e);
+            }
+
+            await SleepAnHour(stoppingToken);
+        }
+    }
+
+    private async Task Claim(Guid subjectId, List<GranularCertificate> consumptionCertificates, List<GranularCertificate> productionCertificates)
+    {
+        foreach (var consumptionCertificate in consumptionCertificates)
+        {
+            var productionCertificate = productionCertificates.FirstOrDefault(x =>
+                x.Start == consumptionCertificate.Start && x.End == consumptionCertificate.End &&
+                x.GridArea == consumptionCertificate.GridArea);
```

Comment:
With this logic a consumption cert will only be used in a single claim, but a production cert can be used multiple times. Why?

Reply:
Good point. For no other reason than that I needed to enumerate one of them. How would you suggest doing it so that the consumption certificate can be used multiple times as well?